TEPHRA-82 Minimize change set retained by client for row-level conflict detection

ghelmling opened the pull request:
This patch minimizes the change set retained in memory by TransactionAwareHTable, where possible.  For HBase 0.94, we still need to retain the full row + family + qualifier for all write in the change set.  However in HBase 0.96+, the delete API now supports a version-specific column family delete.  This allows us to retain just the row + family in the change set when using row-level conflict detection.  In the case of transaction rollback, we simply issue version specific deletes for each row + family that was written.  This minimizes the change set we must retain in memory and reduces the number of delete operations we must issue for rollback, potentially by a large amount.

henry-cask commented on tephra-hbase-compat-0.94/src/main/java/co/cask/tephra/hbase94/TransactionAwareHTable.java:
```diff
@@ -515,6 +506,46 @@ private ActionChange(byte[] row, byte[] family, byte[] qualifier) {
     }
+
+    @Override
+    public boolean equals(Object o) {
```
Maybe add quick optimization to check:
```
if (obj == this) {
  return true;
}
```

after checking for class

ghelmling replied:
Good idea.  On looking at this again, I think we can actually extract a common base class for the TransactionAwareHTable implementations, so we can reduce a lot of the code duplication here.

ghelmling commented:
Updated with a refactored version that pulls all the common parts into an abstract base class -- AbstractTransactionAwareTable -- which all the HBase version-specific implementations now extend.  This removes a lot of the code duplication, though it may make this change a little more difficult to review.

I also added the instance equality check and early out to ActionChange.equals().

henry-cask commented on tephra-hbase-compat-0.94/src/main/java/co/cask/tephra/hbase94/TransactionAwareHTable.java:
```diff
@@ -97,36 +91,72 @@ public TransactionAwareHTable(HTableInterface hTable, boolean allowNonTransactio
+        byte[] qualifier = change.getQualifier();
+        long transactionTimestamp = tx.getWritePointer();
+        Delete rollbackDelete = new Delete(row, transactionTimestamp);
+        if (family != null && qualifier == null) {
```
Not checking the conflictLevel for 94 version?

ghelmling replied:
Yes, that is correct.  HBase 0.94 does not support a version-specific column family delete operation, which is what we use to rollback changes with row-level conflict detection.  So this retains the old behavior by keeping column level change sets for rollback, regardless of the conflict detection level used.

For HBase 0.96+, we only need to retain the row + family for each change, and use family deletes for rollback.

henry-cask commented:
+1 LGTM
